### ISSUE TITLE
Option to disable the `null` strictness with strict-type-predicates

### ISSUE THREAD

**Issue by Jessidhia** [palantir/tslint#2567]
### Bug Report

- __TSLint version__: 5.1.0
- __TypeScript version__: 2.3.0
- __Running TSLint via__: tslint-loader/webpack

#### TypeScript code being linted

```ts
let foo: string|undefined = await externalApiCall()
if (foo != null) {
  // ...
}
```

with `tslint.json` configuration:

```json
{
  "extends": [
    "tslint-config-standard"
  ]
}
```

#### Actual behavior

Requirement that `foo` be compared as `foo !== undefined`.

#### Expected behavior

Optionally, at least, allowing specifically the double-equals comparison with `null`.

The type definitions may not be perfect, especially when dealing with external data, functions that are typed as returning `any`, etc. Requiring that all comparisons with `null` / `undefined` be strict actually makes the code less safe at runtime, especially when there is no particular need to distinguish `null` from `undefined`.

It could be good to always enforce that if the type of the value includes both `null` and `undefined` in its type, at least if the `undefined` is explicit as opposed to implicit from being an optional key/parameter.

**Comment by andy-hanson** [comment]
It already won't warn for checks on a value of type `any`.

In the given example, why not just write `let foo: string | null | undefined`? You're contradicting yourself by typing it as non-null and then checking for null anyway. I don't see how your example is different from `let foo: string | number` and then checking `if (foo !== null) ...`.

The whole point of `strict-type-predicates` is to make up for TypeScript's behavior of allowing testing for values that shouldn't be possible. If you liked that behavior, then you should disable the rule.

**Comment by ypresto** [comment]
(Perhaps this behavior added in below comment)
https://github.com/palantir/tslint/pull/2046#discussion_r96172998

Still, `nullableValue != null` is not always negative.
For example I have used `==` because typescript warns if different types are used (i.e. something like `1 == "1"`).
IMO `triple-equals` rule already implements check for `===`/`!==` and `FAILURE_STRICT_PREFER_STRICT_EQUALS` in `strict-type-predicates` rule is not necessary.

**Comment by ypresto** [comment]
Ah, I see. I want to use this behavior (`Requirement that foo be compared as foo !== undefined.`) when I'm using `triple-equals` with allow-null-check.
So I think it'd be better to be configurable for people who does not use `triple-equals`.

**Comment by jupl** [comment]
This is hurting me right now when I have `strict-type-predicates`, `no-null-keyword`, and `triple-equals` with `allow-undefined-check`. Let's say `x` is of type `string | null` (because of 3rd party library) and I want to verify that `x` is not null:

```ts
x != undefined // Use '!== null' instead. [strict-type-predicates]
x !== null // Use 'undefined' instead of 'null'. [no-null-keyword]
```

**Comment by adidahiya** [comment]
@jupl I think your issue https://github.com/palantir/tslint/issues/2798 tackles this better head-on, and the fix is likely not in the `strict-type-predicates` rule.

@Kovensky I agree that this causes friction when dealing with third-party code, but I do find it helps find errors in declared typings, as @andy-hanson suggests. If you really expect `string | undefined | null` as a return value then you should have it declared as such (fix upstream typings, etc.). At the same time, if we get feedback that the rule is too strict for general use, then we will consider making it more lenient. It's not enabled in tslint's default configs, so we aren't getting a ton of feedback on it yet, but I think we should change this in the next major version (enable some type checking rules in `tslint:recommended`).

For now, I'm inclined to leave the rule as-is.

**Comment by Kovensky** [comment]
Without adding an option this rule is incompatible with https://palantir.github.io/tslint/rules/triple-equals/ 's options, though.

**Comment by adidahiya** [comment]
@Kovensky can you explain that further? Which part is incompatible?

**Comment by Kovensky** [comment]
Thinking more about it, it is incompatible in "spirit" rather than in "letter".

triple-equals has options specifically to allow `!= null` (or `!= undefined`) to check for both cases at the same time because most of the time it doesn't matter if you do triple-equals for those cases, but this rule requires triple-equals even for the few cases in the language where using double equals is "safe".

**Comment by w0rp** [comment]
This is worth supporting, as `x != null` is a safer check to make than `x !== null`. The type system might believe the type of `x` is `T | null`, but `x` might actually become `undefined` at runtime. So a `x != null` check could lead to fewer bugs in your code, if you just want to check if something is is neither null or possibly undefined.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!
